### PR TITLE
Refactor subscriptions code so it can be used in Fusion

### PR DESCRIPTION
**Pull request by roccotripaldi**
Fixes a problem we've seen in developing the subscription block where we sometimes
forget to make the equivalent upstream changes.

#### Changes proposed in this Pull Request:

* Moves the class Jetpack_Subscriptions_Widget into /modules/subscriptions/views.php
* Moves the registration for the subscriptions block into /modules/subscriptions/views.php
* /modules/subscriptions/views.php will be shared with wordpress.com

#### Testing instructions:

- Get this branch running locally or on jurassic.ninja
- Ensure subscription widgets still function as usual
- add a widget to the blog
- add a subscription block to a post
- try subscribing from the front end

#### Proposed changelog entry for your changes:
No changelog needed

### CODE REVIEW

**Review by jeherve — CHANGES_REQUESTED** [review]
I'm seeing some warnings when trying to use the forms on a test site:

```
Strict Standards: Non-static method Jetpack_Subscriptions_Widget::defaults() should not be called statically in wp-content/plugins/jetpack-dev/modules/subscriptions/views.php on line 640

Notice: Undefined index: subscribe_logged_in in wp-content/plugins/jetpack-dev/modules/subscriptions/views.php on line 227
```

**Review by jeherve — CHANGES_REQUESTED** [review]
I added a few more comments.

I think it may be worth looking at the output from `phpcs` in your editor on the new changes; it helps and gives quite a few recommendations.

There is also a mix of spaces added all over the new view file, instead of tabs. Your editor's linter should be able to clean things up a bit.

Overall, it makes me wonder if it would not be easier to get rid of as much duplication as possible, and instead make a few changes to the WordPress.com widget to unify things a bit. We basically have 2 different interfaces in one file right now, that try to do the same thing, and used to do the same thing, but have not been kept in sync so far.

File: modules/subscriptions/views.php [hunk 1]
```diff
@@ -0,0 +1,670 @@
...
+		$name = self::is_jetpack() ?
+			/**
+			 * Filter the name of a widget included in the Extra Sidebar Widgets module.
+			 *
+			 * @module widgets
+			 *
+			 * @since 2.1.2
+			 *
+			 * @param string $widget_name Widget title.
+			 */
```
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: will that make sense to someone browsing this file on the wpcom code base?

File: modules/subscriptions/views.php [hunk 2]
```diff
@@ -0,0 +1,670 @@
...
+		} else {
+			$current_user = wp_get_current_user();
+			if ( ! empty( $current_user->user_email ) ) {
+				$subscribe_email = esc_attr( $current_user->user_email );
+			} else {
+				$subscribe_email = '';
+			}
+		}
+
+		if ( self::is_jetpack() ) {
```
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: Thanks for explaining this!

File: modules/subscriptions/views.php [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+			} else {
+				if ( ! empty( $instance['title'] ) ) {
+					echo $before_title . '<label for="subscribe-field' . ( Jetpack_Subscriptions_Widget::$instance_count > 1 ? '-' . Jetpack_Subscriptions_Widget::$instance_count : '' ) . '">' . esc_attr( $instance['title'] ) . '</label>' . $after_title . "\n";
+				}
+			}
+		}
+
+		if ( self::is_jetpack() && empty( $instance['show_only_email_and_button'] ) ) {
+			echo $args['before_title'] . esc_attr( $instance['title'] ) . $args['after_title'] . "\n";
+		}
```
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: No fall back needed - in the case of Jetpack but with $instance['show_only_email_and_button'] set - we will do nothing.

I'm adding a docblock that should help clear this up.

File: modules/subscriptions/views.php [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+	}
+
+	static function render_widget_title( $instance, $args ) {
+		$show_only_email_and_button = $instance['show_only_email_and_button'];
+		$before_title               = isset( $args['before_title'] ) ? $args['before_title'] : '';
+		$after_title                = isset( $args['after_title'] ) ? $args['after_title'] : '';
+		if ( self::is_wpcom() && ! $show_only_email_and_button ) {
+			if ( self::is_current_user_subscribed() ) {
+				if ( ! empty( $instance['title_following'] ) ) {
+					echo $before_title . '<label for="subscribe-field' . ( Jetpack_Subscriptions_Widget::$instance_count > 1 ? '-' . Jetpack_Subscriptions_Widget::$instance_count : '' ) . '">' . esc_attr( $instance['title_following'] ) . '</label>' . $after_title . "\n";
```
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: I don't know how to fix this or make it more clear.

File: modules/subscriptions/views.php [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+			$show_subscribers_total = (bool) $instance['show_subscribers_total'];
+			if ( $show_subscribers_total && $subscribers_total > 1 ) :
+				$subscribers_not_me = $subscribers_total - 1;
+				?>
+                <p><?php printf( _n( 'You are following this blog, along with %s other amazing person (<a href="%s">manage</a>).', 'You are following this blog, along with %s other amazing people (<a href="%s">manage</a>).', $subscribers_not_me ), number_format_i18n( $subscribers_not_me ), $edit_subs_url ) ?></p><?php
+			else :
+				?>
+                <p><?php printf( __( 'You are following this blog (<a href="%s">manage</a>).' ), $edit_subs_url ) ?></p><?php
+			endif;
+		}
```
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: I'll leave it as is - it's nice to see the difference between the code bases
In wpcom we can easily know if someone is following the blog, so we show a helpful message.
In jetpack, it's not as easy to know, so we do nothing.

File: modules/subscriptions/views.php [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+	 */
+	static function defaults() {
+		$defaults = array(
+			'show_subscribers_total'     => true,
+			'show_only_email_and_button' => false
+		);
+
+		if ( self::is_jetpack() ) {
+			$defaults['title']                 = esc_html__( 'Subscribe to Blog via Email', 'jetpack' );
+			$defaults['subscribe_text']        = esc_html__( 'Subscribe to Blog via Email', 'jetpack' );
```
Comment (jeherve) [inline modules/subscriptions/views.php]: Those default values do not appear to be correct; they are all the same

File: modules/subscriptions/views.php [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+
+	/**
+	 * Enqueue the form's CSS.
+	 *
+	 * @since 4.5.0
+	 */
+	function enqueue_style() {
+		wp_register_style(
+			'jetpack-subscriptions',
+			plugins_url( 'subscriptions/subscriptions.css', __FILE__ ),
```
Comment (jeherve) [inline modules/subscriptions/views.php]: This won't work in Jetpack; it tries to load `wp-content/plugins/jetpack/modules/subscriptions/subscriptions/subscriptions.css?ver=6.9-alpha`

**Review by jeherve — CHANGES_REQUESTED** [review]
I spotted a couple of additional issues that it would be nice to fix as well.

File: modules/subscriptions/views.php (hunk 6 above)
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: great catch!

**Comment by roccotripaldi** [comment]
thanks for the reviews @jeherve - i addressed your feedback in a few recent commits.

today we are going to try to merge the .com side, and if all goes well there, i'd say we are ready to merge this.

see D22567-code

File: modules/subscriptions/views.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dereksmart) [reply modules/subscriptions/views.php]: @roccotripaldi maybe not, but these docblocks are for developer.jetpack.com - writing the comment as Jeremy suggested will ensure there aren't duplicate entries on that site, and it follows the convention of all other filter docblocks.

File: modules/subscriptions/views.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (dereksmart) [reply modules/subscriptions/views.php]: This is fine @jeherve

File: modules/subscriptions/views.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (roccotripaldi) [reply modules/subscriptions/views.php]: ah, thanks for clarifying

**Review by jeherve — APPROVED** [review]
Everything seems to work right now. I think we have the potential to do a lot more to simplify all this in the future and get rid of most of the PHPCS errors that old codebase generates, but as you mentioned it's a good start to bring the 2 codebases together. 👍

**Comment by dereksmart** [comment]
+1, this tests well for me and is a good step in the right direction. @roccotripaldi I left one comment in the phab diff regarding checking with VIP themes before merging to wpcom, but aside from that I feel this is good to go.

**Review by dereksmart — APPROVED** [review]

**Pull request merged**